### ISSUE TITLE
[BUG] `plot_haxby_understand_decoder.py` does not show that nilearn and sklearn coincincde when using  penalty='l1'

### ISSUE THREAD

**Issue by bthirion** [nilearn/nilearn#4695]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Operating system

- [x] Linux
- [ ] Mac
- [ ] Windows

### Operating system version

For example one of the following:
- Linux Ubuntu 22.04

### Python version

- [ ] 3.13
- [ ] 3.12
- [ ] 3.11
- [x] 3.10
- [ ] 3.9

### nilearn version

Dev version

### Expected behavior

scores_nilearn=scores_sklearn

### Current behavior & error messages

Nilearn mean AU-ROC score 0.3446134626690182
Scikit-Learn mean AU-ROC score 0.6557080393885949

<details>
<summary>### Steps and code to reproduce bug</summary>

```python

import numpy as np
from nilearn import datasets
import pandas as pd
from nilearn.image import index_img
from nilearn.maskers import NiftiMasker
from matplotlib import pyplot as plt
from sklearn.preprocessing import LabelBinarizer
from matplotlib.colors import ListedColormap
from sklearn.preprocessing import LabelEncoder
from nilearn._utils.param_validation import adjust_screening_percentile
from nilearn.image import load_img
from sklearn.feature_selection import SelectPercentile, f_classif
from sklearn.linear_model import LogisticRegressionCV
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import LeaveOneGroupOut
from nilearn.decoding import Decoder


haxby_dataset = datasets.fetch_haxby()
fmri_img = haxby_dataset.func[0]
mask_vt = haxby_dataset.mask_vt[0]


behavioral_data = pd.read_csv(haxby_dataset.session_target[0], sep=" ")
labels = behavioral_data["labels"]
labels_mask = labels != "rest"
y = labels[labels_mask]
y = y.to_numpy()
run = behavioral_data["chunks"][labels_mask]
run = run.to_numpy()
fmri_img = index_img(fmri_img, labels_mask)
n_labels = len(np.unique(y))

print(f"{n_labels} labels to predict (y): {np.unique(y)}")
print(f"fMRI data shape (X): {fmri_img.shape}")
print(f"Runs (groups): {np.unique(run)}")

masker = NiftiMasker(mask_img=mask_vt, standardize="zscore_sample")

label_binarizer = LabelBinarizer(pos_label=1, neg_label=-1)
y_binary = label_binarizer.fit_transform(y)

y_binary_ = y_binary.copy()
for col in range(y_binary_.shape[1]):
    y_binary_[np.where(y_binary_[:, col] == 1), col] = col

fig, (ax_binary, ax_multi) = plt.subplots(
    2, gridspec_kw={"height_ratios": [10, 1.5]}, figsize=(12, 2)
)
cmap = ListedColormap(["white"] + list(plt.cm.tab10.colors)[0:n_labels])
binary_plt = ax_binary.imshow(
    y_binary_.T,
    aspect="auto",
    cmap=cmap,
    interpolation="nearest",
    origin="lower",
)
ax_binary.set_xticks([])
ax_binary.set_yticks([])
ax_binary.set_ylabel("One-vs-Others")

# encode the original labels for plotting
label_multi = LabelEncoder()
y_multi = label_multi.fit_transform(y)
y_multi = y_multi.reshape(1, -1)
cmap = ListedColormap(list(plt.cm.tab10.colors)[0:n_labels])
multi_plt = ax_multi.imshow(
    y_multi,
    aspect="auto",
    interpolation="nearest",
    cmap=cmap,
)
ax_multi.set_yticks([])
ax_multi.set_xlabel("Original trial sequence")
cbar = fig.colorbar(multi_plt, ax=[ax_binary, ax_multi])
cbar.set_ticks(np.arange(1 + len(label_multi.classes_)))
cbar.set_ticklabels([*label_multi.classes_, "all others"])
plt.show()

mask_vt_loaded = load_img(mask_vt)
screen_percent = adjust_screening_percentile(20, mask_vt_loaded)
print(f"Adjusted screening percentile: {screen_percent}")

feature_selector = SelectPercentile(f_classif, percentile=int(screen_percent))
classifier = LogisticRegressionCV(
    penalty="l1",
    solver="liblinear",
    Cs=np.geomspace(1e-3, 1e4, 8),
    refit=True,
)

logo_cv = LeaveOneGroupOut()

# Transform fMRI data into a 2D numpy array and standardize it with the masker
X = masker.fit_transform(fmri_img)
print(f"fMRI data shape after masking: {X.shape}")

scores_sklearn = []
for klass in range(n_labels):
    for train, test in logo_cv.split(X, y, groups=run):
        X_train, X_test = X[train], X[test]
        y_train, y_test = y_binary[train, klass], y_binary[test, klass]
        X_train = feature_selector.fit_transform(X_train, y_train)
        X_test = feature_selector.transform(X_test)

        classifier.fit(X_train, y_train)
        pred = classifier.predict_proba(X_test)
        score = roc_auc_score(y_test, pred[:, 1])
        scores_sklearn.append(score)


decoder = Decoder(
    estimator="logistic_l1",
    mask=mask_vt,
    standardize="zscore_sample",
    n_jobs=n_labels,
    cv=logo_cv,
    screening_percentile=20,
    scoring="roc_auc_ovr",
)
decoder.fit(fmri_img, y, groups=run)
scores_nilearn = np.concatenate(list(decoder.cv_scores_.values()))

print("Nilearn mean AU-ROC score", np.mean(scores_nilearn))
print("Scikit-Learn mean AU-ROC score", np.mean(scores_sklearn))


```

</details>

**Comment by Remi-Gau** [comment]
I also realize that this example makes use of a private function: 

```python
from nilearn._utils.param_validation import adjust_screening_percentile
```

I don't there is a workaround: unless we want to make it public?
Probably should mention the fact that we use it for pedagogical reasons but that users probably should not use it in production code.

**Comment by bthirion** [comment]
We this we can simply consider screening_percentile = 100

**Comment by man-shu** [comment]
Would have to look more closely but I think the discrepancy arises from the parameter grid we input to the classifier. See: https://github.com/nilearn/nilearn/blob/main/nilearn%2Fdecoding%2Fdecoder.py#L203

In summary, for L2 penalty, the Cs are more or less hard-coded in nilearn decoder and I used them directly in the sklearn classifier. But for L1 penalty they are determined in a data driven way.

**Comment by bthirion** [comment]
OK, thx ! 
This is interesting.
Anyway, we need to be able to replicate btw nilearn and sklearn...

**Comment by man-shu** [comment]
Even after using the correct `Cs` the discrepancy still exists. So that's not the bigger issue.

The issue is rather that Nilearn decoder sets the score to 0 when all coefficients are 0, irrespective of the scoring metric being used; see https://github.com/nilearn/nilearn/blob/567498d5a6d7766ebebbb6aa2c186786a5ea3534/nilearn/decoding/decoder.py#L404

So when using AU-ROC, if the classifier is not working, the sklearn classifier would have a score of 0.5, and the nilearn decoder would have a score of 0 -- and averaging these scores would cause the discrepancy.

Score comparison for each CV split and class
```python
In [8]: scores_sklearn
Out[8]: {'bottle': [0.5, 0.5, 0.5, 0.5, 0.5, 0.9841269841269841, 0.5, 0.8783068783068784, 0.5, 0.5, 0.5, 0.5], 'cat': [0.9594356261022927, 0.9312169312169312, 0.5, 0.5, 0.5, 0.5, 0.5, 0.8130511463844797, 0.5, 0.8871252204585538, 0.9664902998236332, 0.8871252204585538], 'chair': [0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5], 'face': [0.5, 0.5, 0.5, 0.5, 1.0, 0.5, 0.9911816578483246, 0.5, 0.5, 0.5, 0.5, 0.5], 'house': [0.9964726631393298, 0.998236331569665, 1.0, 1.0, 1.0, 0.998236331569665, 1.0, 0.9911816578483245, 1.0, 1.0, 0.9911816578483246, 1.0], 'scissors': [0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5], 'scrambledpix': [0.9947089947089947, 0.9788359788359788, 0.9523809523809523, 0.9964726631393299, 0.9841269841269842, 0.9947089947089948, 0.9876543209876543, 0.9770723104056437, 0.9841269841269842, 1.0, 0.962962962962963, 1.0], 'shoe': [0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5]}

In [11]: scores_nilearn 
Out[11]: {'bottle': [0, 0, 0, 0, 0, 0.9841269841269841, 0, 0.8747795414462081, 0, 0, 0, 0], 'cat': [0.9594356261022927, 0.9312169312169312, 0, 0, 0, 0, 0, 0.8130511463844797, 0, 0.8871252204585538, 0.9664902998236332, 0.8871252204585538], 'chair': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 'face': [0, 0, 0, 0, 0, 0, 0.9911816578483246, 0, 0, 0, 0, 0], 'house': [0.9964726631393298, 0.998236331569665, 1.0, 1.0, 1.0, 0.998236331569665, 1.0, 0.9911816578483245, 1.0, 1.0, 0.9911816578483246, 1.0], 'scissors': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 'scrambledpix': [0.998236331569665, 0.9788359788359788, 0.9523809523809523, 0.9964726631393299, 0.9841269841269842, 0.9947089947089948, 0.9876543209876543, 0.9770723104056437, 0.9841269841269842, 1.0, 0.962962962962963, 0.998236331569665], 'shoe': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}
```

Updated code:
```python
import numpy as np
from nilearn import datasets
import pandas as pd
from nilearn.image import index_img
from nilearn.maskers import NiftiMasker
from matplotlib import pyplot as plt
from sklearn.preprocessing import LabelBinarizer
from matplotlib.colors import ListedColormap
from sklearn.preprocessing import LabelEncoder
from nilearn._utils.param_validation import adjust_screening_percentile
from nilearn.image import load_img
from sklearn.feature_selection import SelectPercentile, f_classif
from sklearn.linear_model import LogisticRegressionCV
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import LeaveOneGroupOut
from nilearn.decoding import Decoder
from sklearn.svm import l1_min_c


haxby_dataset = datasets.fetch_haxby()
fmri_img = haxby_dataset.func[0]
mask_vt = haxby_dataset.mask_vt[0]


behavioral_data = pd.read_csv(haxby_dataset.session_target[0], sep=" ")
labels = behavioral_data["labels"]
labels_mask = labels != "rest"
y = labels[labels_mask]
y = y.to_numpy()
run = behavioral_data["chunks"][labels_mask]
run = run.to_numpy()
fmri_img = index_img(fmri_img, labels_mask)
n_labels = len(np.unique(y))

print(f"{n_labels} labels to predict (y): {np.unique(y)}")
print(f"fMRI data shape (X): {fmri_img.shape}")
print(f"Runs (groups): {np.unique(run)}")

masker = NiftiMasker(mask_img=mask_vt, standardize="zscore_sample")

label_binarizer = LabelBinarizer(pos_label=1, neg_label=-1)
y_binary = label_binarizer.fit_transform(y)

y_binary_ = y_binary.copy()
for col in range(y_binary_.shape[1]):
    y_binary_[np.where(y_binary_[:, col] == 1), col] = col

fig, (ax_binary, ax_multi) = plt.subplots(
    2, gridspec_kw={"height_ratios": [10, 1.5]}, figsize=(12, 2)
)
cmap = ListedColormap(["white"] + list(plt.cm.tab10.colors)[0:n_labels])
binary_plt = ax_binary.imshow(
    y_binary_.T,
    aspect="auto",
    cmap=cmap,
    interpolation="nearest",
    origin="lower",
)
ax_binary.set_xticks([])
ax_binary.set_yticks([])
ax_binary.set_ylabel("One-vs-Others")

# encode the original labels for plotting
label_multi = LabelEncoder()
y_multi = label_multi.fit_transform(y)
y_multi = y_multi.reshape(1, -1)
cmap = ListedColormap(list(plt.cm.tab10.colors)[0:n_labels])
multi_plt = ax_multi.imshow(
    y_multi,
    aspect="auto",
    interpolation="nearest",
    cmap=cmap,
)
ax_multi.set_yticks([])
ax_multi.set_xlabel("Original trial sequence")
cbar = fig.colorbar(multi_plt, ax=[ax_binary, ax_multi])
cbar.set_ticks(np.arange(1 + len(label_multi.classes_)))
cbar.set_ticklabels([*label_multi.classes_, "all others"])
plt.show()

mask_vt_loaded = load_img(mask_vt)
screen_percent = adjust_screening_percentile(20, mask_vt_loaded)
print(f"Adjusted screening percentile: {screen_percent}")

feature_selector = SelectPercentile(f_classif, percentile=int(screen_percent))
classifier = LogisticRegressionCV(
    penalty="l1",
    solver="liblinear",
    refit=True,
)

logo_cv = LeaveOneGroupOut()

# Transform fMRI data into a 2D numpy array and standardize it with the masker
X = masker.fit_transform(fmri_img)
print(f"fMRI data shape after masking: {X.shape}")

scores_sklearn = {f"{label}": [] for label in label_multi.classes_}
for klass in range(n_labels):
    for train, test in logo_cv.split(X, y, groups=run):
        X_train, X_test = X[train], X[test]
        y_train, y_test = y_binary[train, klass], y_binary[test, klass]
        X_train = feature_selector.fit_transform(X_train, y_train)
        X_test = feature_selector.transform(X_test)

        if classifier.penalty == "l1":
            min_c = l1_min_c(X_train, y_train, loss="log")
        else:
            min_c = 0.5

        param_C = np.geomspace(2e-3, 2e4, 8) * min_c
        print(f"Minimum C: {min_c}")
        print(f"Param C: {param_C}")
        # set Cs as param_C
        classifier.Cs = param_C
        classifier.fit(X_train, y_train)
        pred = classifier.predict_proba(X_test)
        score = roc_auc_score(y_test, pred[:, 1])
        scores_sklearn[label_multi.inverse_transform([klass])[0]].append(score)


decoder = Decoder(
    estimator="logistic_l1",
    mask=mask_vt,
    standardize="zscore_sample",
    n_jobs=n_labels,
    cv=logo_cv,
    screening_percentile=20,
    scoring="roc_auc_ovo",
)
decoder.fit(fmri_img, y, groups=run)
scores_nilearn = decoder.cv_scores_
flat_nilearn_scores = np.concatenate(list(scores_nilearn.values()))
flat_sklearn_scores = np.concatenate(list(scores_sklearn.values()))

print("Nilearn mean AU-ROC score", np.mean(flat_nilearn_scores))
print("Scikit-Learn mean AU-ROC score", np.mean(flat_sklearn_scores))
```
